l2/system-contracts/DefaultAccount.sol

```diff
@@ -72,34 +100,54 @@ contract DefaultAccount is IAccount {
 		_execute(_transaction);
 	}
 
+	/// @notice Method that should be used to initiate a transaction from this account 
+	/// by an external call. This is not mandatory, but should be implemented so that
+	/// it is always possible to execute transaction from L1 for this account.
+	/// @dev This method is basically validate + execute.
+	/// @param _transaction The transaction to execute.
 	function executeTransactionFromOutside(Transaction calldata _transaction) external payable override ignoreNonBootloader {
 		// The account recalculate the hash on its own
 		_validateTransaction(bytes32(0), _transaction);
 		_execute(_transaction);
 	}
 
+	/// @notice Inner method for executing a transaction.
+	/// @param _transaction The transaction to execute.
 	function _execute(Transaction calldata _transaction) internal {
 		address to = address(uint160(_transaction.to));
 		uint256 value = _transaction.reserved[1];
+		require(address(this).balance >= value, "Not enough balance to execute");
+
 		bytes memory data = _transaction.data;
 
 		if(to == address(DEPLOYER_SYSTEM_CONTRACT)) {
-			// We allow calling ContractDeployer with any calldata
+			// Note, that the deployer contract can only be called 
+			// with a "systemCall" flag.
 			SystemContractsCaller.systemCall(
 				uint32(gasleft()),
 				to,
 				uint128(_transaction.reserved[1]), // By convention, reserved[1] is `value`
 				_transaction.data
 			);
 		} else {
-			bool success;
 			assembly {
-				success := call(gas(), to, value, add(data, 0x20), mload(data), 0, 0)
+				let success := call(gas(), to, value, add(data, 0x20), mload(data), 0, 0)
+
+				// The returned revertReason will be available on the server
+				// side as the revert reason of the transaction.
+				if iszero(success) {
+					let size := returndatasize()
+					returndatacopy(0, 0, size)
+					revert(0, size)
+				}
```

Comment:
if we're copying over the whole `returndata`, is assembly still needed here?

Reply:
Since we do a low-level assembly `call`, then it is just the most convenient way to return data

Reply:
sure, but why is assembly needed in the first place instead of something like `to.call{value: value}(data)`?

Reply:
I'm trying to withdraw a bunch of contracts do what you thinks best. Maybe you can help me do this correct? I'm new at this.